lib/getWorkflow.js

```diff
@@ -9,80 +9,203 @@
 const filterNodeName = name =>
     (/^~(pr|commit|release|tag|sd@)/.test(name) ? name : name.replace('~', ''));
 
+/**
+ * Build external nodes for its downstream jobs, DFS to traverse its children
+ * @method buildExternalNodes
+ * @param  {String}         root           Source name. ex `sd@123:main`
+ * @param  {Array}          children       Array of downstream jobs. ex: [sd@456:test, sd@789:test]
+ * @param  {Set}            nodes          List of graph nodes
+ * @param  {TriggerFactory} triggerFactory triggerFactory to figure out its downstream jobs
+ * @return {Promise}  List of graph nodes
+ */
+const buildExternalNodes = async (root, children, nodes, triggerFactory) => {
+    await Promise.all(children.map(async (dest) => {
+        nodes.add(dest);
+
+        const newRoot = dest;
+        const newChildren = await triggerFactory.getDestFromSrc(newRoot);
+
+        await buildExternalNodes(newRoot, newChildren, nodes, triggerFactory);
+    }));
+};
+
 /**
  * Get the list of nodes for the graph
  * @method calculateNodes
- * @param  {Object}       jobs Hash of job configs
- * @return {Array}             List of nodes (jobs)
+ * @param  {Object}          jobs           Hash of job configs
+ * @param  {TriggerFactory}  triggerFactory Trigger Factory to find external triggers
+ * @param  {Number}          pipelineId     Id of the current pipeline
+ * @return {Array}           List of nodes (jobs)
  */
-const calculateNodes = (jobs) => {
+const calculateNodes = async (jobs, triggerFactory, pipelineId) => {
     const nodes = new Set(['~pr', '~commit']);
 
-    Object.keys(jobs).forEach((name) => {
-        nodes.add(name);
-        if (Array.isArray(jobs[name].requires)) {
-            jobs[name].requires.forEach(n => nodes.add(filterNodeName(n)));
+    // for backward compatibility. TODO: remove this block later
+    if (!triggerFactory) {
+        Object.keys(jobs).forEach((name) => {
+            nodes.add(name);
+            if (Array.isArray(jobs[name].requires)) {
+                jobs[name].requires.forEach(n => nodes.add(filterNodeName(n)));
+            }
+        });
+
+        return [...nodes].map(name => ({ name }));
+    }
+
+    // new implementation. allow external join
+    await Promise.all(Object.keys(jobs).map(async (jobName) => {
+        nodes.add(jobName);
+
+        // upstream nodes
+        if (Array.isArray(jobs[jobName].requires)) {
+            jobs[jobName].requires.forEach(n => nodes.add(filterNodeName(n)));
         }
-    });
+
+        // downstream nodes
+        const srcName = `sd@${pipelineId}:${jobName}`;
+
+        // Calculate which downstream jobs are triggered BY current job
+        // Only need to take care of external triggers, since internal will be taken care automatically
+        const externalDownstreamOr = await triggerFactory.getDestFromSrc(`~${srcName}`);
```

Comment:
where is the `getDestFromSrc` function defined?

Reply:
Good catch. It needs to be added to this PR: https://github.com/screwdriver-cd/models/pull/400/files

Reply:
added ✅